accounts/abi/bind/base.go

```diff
@@ -0,0 +1,178 @@
+	caller     ContractCaller     // Read interface to interact with the blockchain
+	transactor ContractTransactor // Write interface to interact with the blockchain
+}
+
+// NewBoundContract creates a low level contract interface through which calls
+// and transactions may be made through.
+func NewBoundContract(address common.Address, abi abi.ABI, caller ContractCaller, transactor ContractTransactor) *BoundContract {
+	return &BoundContract{
+		address:    address,
+		abi:        abi,
+		caller:     caller,
+		transactor: transactor,
+	}
+}
+
+// DeployContract deploys a contract onto the Ethereum blockchain and binds the
+// deployment address with a Go wrapper.
+func DeployContract(opts *TransactOpts, abi abi.ABI, bytecode []byte, backend ContractBackend) (*BoundContract, error) {
+	// Sanity check the authorization options
+	if opts == nil {
+		return nil, errors.New("transaction options missing")
+	}
+	// Otherwise try to deploy the contract
+	c := NewBoundContract(common.Address{}, abi, backend.(ContractCaller), backend.(ContractTransactor))
+
+	tx, err := c.transact(opts, nil, bytecode)
+	if err != nil {
+		return nil, err
+	}
+	c.address = crypto.CreateAddress(opts.Account, tx.Nonce())
+	return c, nil
+}
+
+// Call invokes the (constant) contract method with params as input values and
+// sets the output to result. The result type might be a single field for simple
+// returns, a slice of interfaces for anonymous returns and a struct for named
+// returns.
+func (c *BoundContract) Call(opts *CallOpts, result interface{}, method string, params ...interface{}) error {
+	// Don't crash on a lazy user
+	if opts == nil {
+		opts = new(CallOpts)
+	}
+	// Pack the input, call and unpack the results
+	input, err := c.abi.Pack(method, params...)
+	if err != nil {
+		return err
+	}
+	output, err := c.caller.ContractCall(c.address, input, opts.Pending)
+	if err != nil {
+		return err
+	}
+	return c.abi.Unpack(result, method, output)
+}
+
+// Transact invokes the (paid) contract method with params as input values and
+// value as the fund transfer to the contract.
+func (c *BoundContract) Transact(opts *TransactOpts, method string, params ...interface{}) (*types.Transaction, error) {
+	// Sanity check the authorization options
+	if opts == nil {
+		return nil, errors.New("transaction options missing")
+	}
+	// Otherwise pack up the parameters and invoke the contract
+	input, err := c.abi.Pack(method, params...)
+	if err != nil {
+		return nil, err
+	}
+	return c.transact(opts, &c.address, input)
+}
+
+// transact executes an actual transaction invocation, first deriving any missing
+// authorization fields, and then scheduling the transaction for execution.
+func (c *BoundContract) transact(opts *TransactOpts, contract *common.Address, input []byte) (*types.Transaction, error) {
+	var err error
+
+	// Ensure a valid value field and resolve the account nonce
+	value := opts.Value
+	if value == nil {
+		value = new(big.Int)
+	}
+	nonce := uint64(0)
```

Comment:
Could you place a comment here saying something like "If no nonce was specified we'll attempt to guess one using ... etc".